src/Toolchain.ml

```diff
@@ -0,0 +1,500 @@
+open Bindings
+open Utils
+module R = Result
+
+type commandAndArgs = string * string array
+
+let promptSetup ~f =
+  let open Js.Promise in
+  (Window.showQuickPick [| "yes"; "no" |]
+     (Window.QuickPickOptions.make ~canPickMany:false
+        ~placeHolder:{j|Setup this project's toolchain with 'esy'?|j} ()) [@bs])
+  |> Js.Promise.then_ (fun choice ->
+         match Js.Nullable.toOption choice with
+         | None -> resolve (Error "Please setup the toolchain")
+         | Some choice ->
+           if choice = "yes" then
+             f ()
+           else
+             resolve (Error "Please setup the toolchain"))
+
+let setupWithProgressIndicator esyCmd ~envWithUnzip:esyEnv folder =
+  let open Setup.Bsb in
+  Window.withProgress
+    [%bs.obj { location = 15; title = "Setting up toolchain..." }]
+    (fun progress ->
+      let succeeded = ref (Ok ()) in
+      let eventEmitter = make () in
+      onProgress eventEmitter (fun percent ->
+          Js.log2 "Percentage:" percent;
+          (progress.report
+             [%bs.obj { increment = int_of_float (percent *. 100.) }] [@bs]));
+      onEnd eventEmitter (fun () ->
+          (progress.report [%bs.obj { increment = 100 }] [@bs]));
+      onError eventEmitter (fun errorMsg -> succeeded := Error errorMsg);
+      let open Js.Promise in
+      run esyCmd esyEnv eventEmitter folder
+      |> then_ (fun () -> resolve !succeeded))
+
+module Binaries = struct
+  let esy = "esy"
+
+  let opam = "opam"
+end
+
+module PackageManager : sig
+  type t
+
+  type spec
+
+  val ofName : Fpath.t -> string -> (t, string) result
+
+  val toName : t -> string
+
+  val toString : t -> string
+
+  val specOfName :
+       env:string Js.Dict.t
+    -> name:string
+    -> root:Fpath.t
+    -> (spec, string) result Js.Promise.t
+
+  val make : env:string Js.Dict.t -> t:t -> (spec, string) result Js.Promise.t
+
+  val setupToolChain : Fpath.t -> spec -> (unit, string) result Js.Promise.t
+
+  val available :
+    env:string Js.Dict.t -> root:Fpath.t -> (t list, string) result Js.Promise.t
+
+  val env : spec -> (string Js.Dict.t, string) result Js.Promise.t
+
+  val lsp : spec -> commandAndArgs
+
+  val compare : t -> t -> int
+
+  val find : string -> t list -> t option
+
+  val esy : Fpath.t -> t
+
+  val opam : Fpath.t -> t
+end = struct
+  type t =
+    | Opam of Fpath.t
+    | Esy of Fpath.t
+    | Global
+
+  type spec =
+    { cmd : Cmd.t
+    ; t : t
+    }
+
+  let esy root = Esy root
+
+  let opam root = Opam root
+
+  let make ~env ~t =
```

Comment:
1. I'd expect from `make` function to produce `t`. Instead it takes `t` and does something with it. I think this function needs a better name. 
2. `t` is a very cryptic label for an argument. As @rgrinberg mentioned earlier, labels should explain arguments, here it doesn't help. But I think with the better function name you will not need label at all. 

`t` is kinda cryptic. We either need a more descriptive name, or no name at all if the

Reply:
Renamed this function to `makeSpec`